src/assets/wise5/classroomMonitor/dataExport/strategies/AbstractComponentDataExportStrategy.ts

```diff
@@ -0,0 +1,367 @@
+      'Project Name',
+      'Run ID',
+      'Start Date',
+      'End Date',
+      'Server Timestamp',
+      'Client Timestamp',
+      'Node ID',
+      'Component ID',
+      'Component Part Number',
+      'Step Title',
+      'Component Type',
+      'Component Prompt',
+      'Student Data',
+      'Component Revision Counter',
+      'Is Submit',
+      'Submit Count'
+    ];
+    for (let c = 0; c < defaultColumnNames.length; c++) {
+      const defaultColumnName = defaultColumnNames[c];
+      columnNameToNumber.set(defaultColumnName, c);
+      columnNames.push(defaultColumnName);
+    }
+  }
+
+  generateComponentWorkRows(component: Component, nodeId: string): string[] {
+    let rows = [];
+    let rowCounter = 1;
+    for (const workgroupId of this.configService.getClassmateWorkgroupIds()) {
+      const rowsForWorkgroup = this.generateComponentWorkRowsForWorkgroup(
+        workgroupId,
+        nodeId,
+        component.id,
+        rowCounter
+      );
+      rows = rows.concat(rowsForWorkgroup);
+      rowCounter += rowsForWorkgroup.length;
+    }
+    return rows;
+  }
+
+  private generateComponentWorkRowsForWorkgroup(
+    workgroupId: number,
+    nodeId: string,
+    componentId: string,
+    rowCounter: number
+  ): string[] {
+    const rows = [];
+    const userInfo = this.configService.getUserInfoByWorkgroupId(workgroupId);
+    const extractedUserIDsAndStudentNames = this.extractUserIDsAndStudentNames(userInfo.users);
+    const componentRevisionCounter = new ComponentRevisionCounter();
+    const componentStates = this.teacherDataService.getComponentStatesByWorkgroupIdAndComponentId(
+      workgroupId,
+      componentId
+    );
+    for (let c = 0; c < componentStates.length; c++) {
+      const componentState = componentStates[c];
+      if (this.shouldExportRow(componentState, c, componentStates.length)) {
+        const row = this.generateComponentWorkRow(
+          rowCounter,
+          workgroupId,
+          extractedUserIDsAndStudentNames['userId1'],
+          extractedUserIDsAndStudentNames['userId2'],
+          extractedUserIDsAndStudentNames['userId3'],
+          extractedUserIDsAndStudentNames['studentName1'],
+          extractedUserIDsAndStudentNames['studentName2'],
+          extractedUserIDsAndStudentNames['studentName3'],
+          userInfo.periodName,
+          componentRevisionCounter,
+          componentState
+        );
+        rows.push(row);
+        rowCounter++;
+      } else {
+        this.incrementRevisionCounter(componentRevisionCounter, nodeId, componentId);
+      }
+    }
+    return rows;
+  }
+
+  /**
+   * @param users An array of user objects. Each user object contains an id and name.
+   * @returns {object} An object that contains key/value pairs. The key is userIdX
+   * or studentNameX where X is an integer. The values are the corresponding actual
+   * values of user id and student name.
+   */
+  private extractUserIDsAndStudentNames(users: any[]): any {
```

Comment:
Define class to store this object that ```extends Map<string, string>```?
```suggestion
  private extractUserIDsAndStudentNames(users: any[]): UserIdsAndStudentNames {
```